File: gfg/2024_6_june/c++/11_maximum_tip_calculator.cpp
// TIME COMPLEXITY
// O(n)

// SPACE COMPLEXITY
// O(n)

// CODE

class Solution {
  public:
    long long maxTip(int n, int x, int y, vector<int> &arr, vector<int> &brr) {
        vector<pair<int,pair<int,int>>> diffArr;
        
        for(int i = 0; i < arr.size(); i++){
            diffArr.push_back({abs(arr[i] - brr[i]), {arr[i], brr[i]}});
        }
        
        sort(diffArr.rbegin(), diffArr.rend());
        
        long long ans = 0;
        
        for(int i = 0; i < diffArr.size(); i++){
            if(x == 0){
                ans += diffArr[i].second.second;
            }
            else if(y == 0){
                ans += diffArr[i].second.first;
            }
            else{
                if(diffArr[i].second.first > diffArr[i].second.second){
                    x--;
                    ans += diffArr[i].second.first;
                }
                else{
                    y--;
                    ans += diffArr[i].second.second;
                }
            }
        }
        
        return ans;
        
    }
};

File: gfg/2024_6_june/python/11_maximum_tip_calculator.py
# CODE

from typing import List
class Solution:
    def maxTip(self, n : int, x : int, y : int, arr : List[int], brr : List[int]) -> int:
        tips = zip(arr, brr)
        tips_sorted = sorted(tips, key = lambda t: abs(t[0] - t[1]), reverse = True)
        cntA, cntB, res = 0, 0, 0
        for i, j in tips_sorted:
            if (i >= j and cntA < x) or (cntB == y):
                res += i
                cntA += 1
            else:
                res += j
                cntB += 1
        return res